Add support for MiniMax's MiniMax-Text-01

geetu040 opened the pull request:
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/35710

This PR adds [MiniMaxAI](https://www.minimaxi.com/en)'s MiniMax-Text-01 model to Hugging Face Transformers.

- MiniMax-Text-01 is a powerful language model with 456 billion total parameters, of which 45.9 billion are activated per token.
- To better unlock the long context capabilities of the model, MiniMax-Text-01 adopts a hybrid architecture that combines Lightning Attention, Softmax Attention and Mixture-of-Experts (MoE).
- MiniMax-Text-01's training context length is extended to 1 million tokens, and it can handle a context of up to 4 million tokens during the inference.
- On various academic benchmarks, MiniMax-Text-01 also demonstrates the performance of a top-tier model.

Relevant Links
- Research Paper: [MiniMax-01: Scaling Foundation Models with Lightning Attention](https://arxiv.org/abs/2501.08313)
- Authors: [MiniMax](https://arxiv.org/search/cs?searchtype=author&query=MiniMax), [Aonian Li](https://arxiv.org/search/cs?searchtype=author&query=Li,+A), [Bangwei Gong](https://arxiv.org/search/cs?searchtype=author&query=Gong,+B), et al.
- Implementation: [MiniMaxAI/MiniMax-Text-01](https://huggingface.co/MiniMaxAI/MiniMax-Text-01/tree/main)
- Models Weights: [MiniMaxAI/MiniMax-Text-01](https://huggingface.co/MiniMaxAI/MiniMax-Text-01)
- Github Page: [MiniMax-AI/MiniMax-01](https://github.com/MiniMax-AI/MiniMax-01)

CC: @MiniMax-AI-Dev

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker, @Rocketknight1

## Change Log

- **Tokenizer**: It uses the existing `GPT2Tokenizer`
- **Config**: Matches the `MixtralConfig` with a few additional parameters:
  - residual_post_norm, attn_type_list
  - layernorm_attention_alpha, layernorm_lightning_attention_alpha, layernorm_mlp_alpha
  - layernorm_attention_beta, layernorm_lightning_attention_beta, layernorm_mlp_beta
- **Weight Conversion Script**: No script needed, original weights can be loaded directly into the new architecutre
- **Model**: `MiniMax-Text-01` architecture matches and uses most of the `Mixtral` architecture, with a few changes in
  - **DecoderLayer**
    - `hidden_states` can be used as residual connections, before or after `layernorm` is applied
    - weighted sum is used in residual connection
    - selection between softmax and lightning attention based on the `layer_idx`
  - **LightningAttention**
    - intially used in [TransNormerLLM](https://arxiv.org/abs/2307.14995), upgraded in [Lightning Attention-2](https://arxiv.org/abs/2401.04658) and adopted in [MiniMax-01](https://arxiv.org/abs/2501.08313)
    - every 8th decoder layer uses a softmax attention, rest of the layers use lightning attention, which is not previously implemented in `transformers`

**To summarize above, the main area of review is the `LightningAttention` implementation.**

## TODOs

- [x] Update Documentation
- [x] Update Tests
- [x] Import Statements and Auto Modeling
- [x] Implement Model
  - [x] Implement End-to-End Architecture
  - [x] Implement `LightningAttention`
    - [x] Work with avialable code
    - [x] Refactor, Clean and Optimize
    - [x] Implement Decays
    - [x] Implement Caching
    - [x] Implement attention_mask
    - [x] Support `.generate()` method
- [x] Fix CI/CD tests
- [x] [MiniMax-Tiny](https://huggingface.co/geetu040/MiniMax-tiny) for slow integration tests
- [ ] Model Card (only config and README file is to be updated)

geetu040 committed:
end-to-end architecture

Rocketknight1 commented:
Hi @geetu040, this looks quite good! You can ping us whenever it's ready for review. Also, code quality issues in the CI can be fixed with `pip install -e .[quality]` in the `transformers` directory, followed by `make fixup`.

geetu040 committed:
lightning-attn: refactor, clean, optimize

geetu040 committed:
put minimax_text_01 in other files

geetu040 committed:
use latest __init__ standards and auto-generate modular

geetu040 committed:
support attention_mask for lightning-attn

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
Revert "use latest __init__ standards and auto-generate modular"

This reverts commit d8d3c409d89e335c98a8cd36f47304a76eac7493.

geetu040 committed:
fix modular conversion

geetu040 committed:
pass both attention masks instead of tuple

geetu040 committed:
formatting

ArthurZucker commented:
Eager to see this! 🔥

Shakib-IO committed:
Updated Dynamic Cache

geetu040 committed:
created MiniMaxText01Cache

geetu040 committed:
fix hardcoded slope_rate

geetu040 committed:
update attn_type_list in config

geetu040 committed:
fix lightning when use_cache=False

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
copy tests from mixtral

geetu040 committed:
(checkpoint) all tests pass for normal attention

geetu040 committed:
fix all unittests

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
fix import sorting

geetu040 committed:
fix consistency and formatting tests

ArthurZucker commented:
cc @Cyrilvallez can you have a look!

geetu040 committed:
fix config

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
update tests, since changes in main

geetu040 committed:
fix seq_len error

geetu040 committed:
create dummy docs

geetu040 committed:
fix checkpoint

geetu040 committed:
add checkpoint in config docstring

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
run modular_conversion

geetu040 committed:
update docs

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
fix checkpoint path and update tests

geetu040 committed:
fix ruff

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
remove repeated expected_slice

geetu040 committed:
update docs

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 commented:
@ArthurZucker, @Rocketknight1, @Cyrilvallez

This PR is ready for review.

Please see `MiniMaxText01Cache`, the lightning attention uses a cache that doesnot grow with the sequence length. Though I have implemented this in the best of my knowledge, still I am not sure if this cache can be traced back for beam search or low memory contrasitive search where we need to crop back the cache given the sequence

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+class MiniMaxText01Cache(DynamicCache):
+    def __init__(self, config, batch_size, dtype=torch.float16, device=None):
+        super().__init__()
+        self.config = config
+        self.num_hidden_layers = config.num_hidden_layers
+        self.attn_type_list = config.attn_type_list
+        self.head_dim = getattr(config, "head_dim", config.hidden_size // config.num_attention_heads)
+        self.num_heads = config.num_attention_heads
```
This looks like it should not be needed

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+    @classmethod
+    def from_legacy_cache(cls, past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None) -> "DynamicCache":
+        raise NotImplementedError("MiniMaxText01Cache does not have a legacy cache equivalent.")
+
+    def __len__(self):
+        """
+        Support for backwards-compatible `past_key_value` length, e.g. `len(past_key_value)`. This value corresponds
+        to the number of layers in the model.
+        """
+        return self.num_hidden_layers
+
+    def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
+        """Returns the sequence length of the cached states. A layer index can be optionally passed."""
+        # TODO: deprecate this function in favor of `cache_position`
+        is_empty_layer = (
+            len(self.key_cache) == 0  # no cache in any layer
+            or len(self.key_cache) <= layer_idx  # skipped `layer_idx` and hasn't run a layer with cache after it
+            or len(self.key_cache[layer_idx]) == 0  # the layer has no cache
+            or self.key_cache[layer_idx].dim() == 2  # specific to MiniMaxText01
+        )
+        layer_seq_length = self.key_cache[layer_idx].shape[-2] if not is_empty_layer else 0
+        return layer_seq_length
+
+    def batch_repeat_interleave(self, repeats: int):
+        """Repeat the cache `repeats` times in the batch dimension. Used in contrastive search."""
+        for layer_idx in range(len(self)):
+            if self.attn_type_list[layer_idx] == 0:
+                self.kv_cache[layer_idx] = self.kv_cache[layer_idx].repeat_interleave(repeats, dim=0)
+            else:
+                self.key_cache[layer_idx] = self.key_cache[layer_idx].repeat_interleave(repeats, dim=0)
+                self.value_cache[layer_idx] = self.value_cache[layer_idx].repeat_interleave(repeats, dim=0)
+
+    def batch_select_indices(self, indices: torch.Tensor):
+        """Only keep the `indices` in the batch dimension of the cache. Used in contrastive search."""
+        for layer_idx in range(len(self)):
+            if self.attn_type_list[layer_idx] == 0:
+                self.kv_cache[layer_idx] = self.kv_cache[layer_idx][indices, ...]
+            else:
+                self.key_cache[layer_idx] = self.key_cache[layer_idx][indices, ...]
+                self.value_cache[layer_idx] = self.value_cache[layer_idx][indices, ...]
```
Looks like all of that is not needed, except maybe `get_seq_length`

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        self.head_dim = getattr(config, "head_dim", config.hidden_size // config.num_attention_heads)
+        self.num_heads = config.num_attention_heads
+
+        self.batch_size = batch_size
+        self.dtype = dtype
+        self.device = device
+
+        self.kv_cache = [torch.tensor([[]] * batch_size, device=device) for _ in range(config.num_hidden_layers)]
+        self.key_cache = [torch.tensor([[]] * batch_size, device=device) for _ in range(config.num_hidden_layers)]
+        self.value_cache = [torch.tensor([[]] * batch_size, device=device) for _ in range(config.num_hidden_layers)]
```
Here, this is extremely hard to read and most of it is redundant/not used. Let's use the layer pattern contained in the config to correctly initialize what we are going to use. Even better, you can create per-layer primitives such as `DynamicCacheLayer` and `LinearCacheLayer`, and then create a list of correct per-layer cache based on the config

geetu040 replied:
I did not create per-layer primitives such as `DynamicCacheLayer` and `LinearCacheLayer`, but have refactored the classes to remove this redundant code

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        self.key_cache = [torch.tensor([[]] * batch_size, device=device) for _ in range(config.num_hidden_layers)]
+        self.value_cache = [torch.tensor([[]] * batch_size, device=device) for _ in range(config.num_hidden_layers)]
+
+    def set_kv_cache(self, kv_cache, layer_idx):
+        self.kv_cache[layer_idx] = kv_cache
+
+    def get_kv_cache(self, layer_idx):
+        return self.kv_cache[layer_idx]
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs=None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Update the number of seen tokens
+        if layer_idx == 0:
+            self._seen_tokens += key_states.shape[-2]
+
+        # Update the cache
+        if self.key_cache[layer_idx].dim() == 2:
+            self.key_cache[layer_idx] = key_states
+            self.value_cache[layer_idx] = value_states
+        else:
+            self.key_cache[layer_idx] = torch.cat([self.key_cache[layer_idx], key_states], dim=-2)
+            self.value_cache[layer_idx] = torch.cat([self.value_cache[layer_idx], value_states], dim=-2)
+
+        return self.key_cache[layer_idx], self.value_cache[layer_idx]
```
Then this should be adapted based on above comment

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        kv_cache = None
+        if past_key_value is not None:
+            kv_cache = past_key_value.get_kv_cache(self.layer_idx)
+
+        if kv_cache is None or kv_cache.dim() == 2:
```
I don't get the `dim == 2` check

geetu040 replied:
removed in current refactor

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        query_states = query_states.transpose(1, 2)
+        key_states = key_states.transpose(1, 2)
+        value_states = value_states.transpose(1, 2)
+
+        kv_cache = None
+        if past_key_value is not None:
+            kv_cache = past_key_value.get_kv_cache(self.layer_idx)
```
Let's not mix `past_key_value` and `kv_cache` here -> `past_key_value` everywhere, or a full other name if it has not link at all to past kv

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+            ratio = torch.exp(-slope_rate)
+            attn_output = []
+            for i in range(seq_len):
+                kv_cache = ratio * kv_cache + torch.einsum(
+                    "... n d, ... n e -> ... d e",
+                    key_states[:, :, i : i + 1],
+                    value_states[:, :, i : i + 1],
+                )
+                attn_output_i = torch.einsum("... n e, ... e d -> ... n d", query_states[:, :, i : i + 1], kv_cache)
```
We really don't like `einsum` in Transformers! Let's use more classical view/transpose operations 🤗

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        output_attentions: Optional[bool] = False,
+        output_router_logits: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # necessary, but kept here for BC
```
Let's make the `position_embeddings` necesary here!

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+
+        self.layer_idx = layer_idx
+        self.postnorm = config.postnorm
+        self.layernorm_full_attention_alpha = config.layernorm_full_attention_alpha
+        self.layernorm_full_attention_beta = config.layernorm_full_attention_beta
+        self.layernorm_linear_attention_alpha = config.layernorm_linear_attention_alpha
+        self.layernorm_linear_attention_beta = config.layernorm_linear_attention_beta
```
Here those 4 declarations of attributes are useless, let's remove them for clarity!

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        self.layernorm_full_attention_beta = config.layernorm_full_attention_beta
+        self.layernorm_linear_attention_alpha = config.layernorm_linear_attention_alpha
+        self.layernorm_linear_attention_beta = config.layernorm_linear_attention_beta
+        self.layernorm_mlp_alpha = config.layernorm_mlp_alpha
+        self.layernorm_mlp_beta = config.layernorm_mlp_beta
```
These 2 parameters (and the other 2), I'm not a huge fan of calling them `layernorm_xxx`, it makes it look like they are Norm layers. Let's use `mlp_alpha/beta_factor` and `attn_alpha/beta_factor` instead! (or something similar)

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+
+        return outputs
+
+
+MINIMAX_TEXT_01_INPUTS_DOCSTRING = MIXTRAL_INPUTS_DOCSTRING
```
You can remove that, it will be picked up automatically!

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+        self.layers = nn.ModuleList(
+            [MiniMaxText01DecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+
+    @add_start_docstrings_to_model_forward(MINIMAX_TEXT_01_INPUTS_DOCSTRING)
```
The decorator should be similarly removed, modualr will add it automatically!

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+                batch_size=batch_size,
+                dtype=dtype,
+            )
+        elif use_cache and not isinstance(past_key_values, MiniMaxText01Cache):
+            past_key_values = MiniMaxText01Cache.from_dynamic_cache(
+                past_key_values,
+                config=config,
+                batch_size=batch_size,
+                dtype=dtype,
+            )
```
Let's not add any equivalence, as there are none if I understand correctly

Cyrilvallez commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+            return_dict=return_dict,
+            cache_position=cache_position,
+            **kwargs,
+        )
+
+        hidden_states = outputs[0]
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
+        logits = self.lm_head(hidden_states[:, slice_indices, :])
+
+        loss = None
+        if labels is not None:
+            loss = self.loss_function(logits, labels, self.vocab_size, **kwargs)
+
+        aux_loss = None
+        if output_router_logits:
+            aux_loss = load_balancing_loss_func(
+                outputs.router_logits if return_dict else outputs[-1],
+                self.num_experts,
+                self.num_experts_per_tok,
+                attention_mask,
+            )
+            if labels is not None:
+                loss += self.router_aux_loss_coef * aux_loss.to(loss.device)  # make sure to reside in the same device
+
+        if not return_dict:
+            output = (logits,) + outputs[1:]
+            if output_router_logits:
+                output = (aux_loss,) + output
+            return (loss,) + output if loss is not None else output
+
+        return MoeCausalLMOutputWithPast(
+            loss=loss,
+            aux_loss=aux_loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            router_logits=outputs.router_logits,
+        )
```
Here this is all similar to Mixtral as well!
```suggestion
    pass
```

Cyrilvallez commented on tests/models/minimax_text_01/test_image_processing_minimax_text_01.py:
Should be removed

Cyrilvallez commented on tests/repo_utils/modular/test_conversion_order.py:
```diff
@@ -19,6 +19,7 @@
     os.path.join(MODEL_ROOT, "granite", "modular_granite.py"),
     os.path.join(MODEL_ROOT, "gemma2", "modular_gemma2.py"),
     os.path.join(MODEL_ROOT, "mixtral", "modular_mixtral.py"),
+    os.path.join(MODEL_ROOT, "minimax_text_01", "modular_minimax_text_01.py"),
     os.path.join(MODEL_ROOT, "olmo", "modular_olmo.py"),
```
This file should not be touched

Cyrilvallez commented on utils/not_doctested.txt:
```diff
@@ -163,6 +163,7 @@ docs/source/en/model_doc/mega.md
 docs/source/en/model_doc/megatron-bert.md
 docs/source/en/model_doc/megatron_gpt2.md
 docs/source/en/model_doc/mgp-str.md
+docs/source/en/model_doc/minimax_text_01.md
 docs/source/en/model_doc/mistral.md
```
This file should not be modified either

Cyrilvallez commented on docs/source/en/conversations.md:
```diff
@@ -281,7 +281,7 @@ confirm these generations with the chat model. If the guesses are validated by t
 
 Finally, we should also note the impact of "Mixture of Experts" (MoE) models here. Several popular chat models,
-such as Mixtral, Qwen-MoE and DBRX, are MoE models. In these models, not every parameter is active for every token generated.
+such as Mixtral, MiniMaxText01, Qwen-MoE and DBRX, are MoE models. In these models, not every parameter is active for every token generated.
```
Humm, it's okay to modify that, but please add your model name at the end of the list then 🤗 And for all languages

Cyrilvallez reviewed (commented):
Hey! Here is a first review! Very nice first modular, congrats! 🤗
Biggest point to work on currently is the Cache class, see my comments 👌
But should not be that much trouble, especially as the modular part itself is quite nice.

I must say I would prefer all classes and files to be named `MiniMax` and `mini_max` however, `MiniMaxText01` feels weird to me. Are you one of the model creator? Do you feel strongly about it?

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
rename "minimax-text-01" to "minimax"

geetu040 committed:
inherit config from mixtral

geetu040 committed:
remove from docs in other languages

geetu040 committed:
undo files that should be untouched

geetu040 committed:
move minimax to end in conversation docs

geetu040 committed:
use MiniMaxForCausalLM as it is

geetu040 committed:
ruff fixes

geetu040 committed:
Merge branch "origin/main"; resolve conflicts

geetu040 committed:
run modular

geetu040 committed:
fix docstring example in causallm

geetu040 committed:
refactor attention loop and decay factors

geetu040 committed:
refactor config in modular

geetu040 committed:
run modular

geetu040 committed:
refactor cache

geetu040 committed:
Merge remote-tracking branch 'origin/main' into minimax-text-01

geetu040 committed:
rename static_cache to linear_cache

geetu040 committed:
make positional embeddings necessary

geetu040 committed:
remove unnecessary layernorms declarations

geetu040 committed:
fix import in tests

geetu040 committed:
refactor attention in next tokens

geetu040 committed:
remove outdated code

geetu040 committed:
formatting and modular

geetu040 commented on src/transformers/models/minimax_text_01/modular_minimax_text_01.py:
```diff
@@ -0,0 +1,887 @@
+
+
+class MiniMaxText01Cache(DynamicCache):
+    def __init__(self, config, batch_size, dtype=torch.float16, device=None):
```
the new refactored class, uses most of parent methods

geetu040 committed:
update tests

geetu040 committed:
rename layernorm alpha/beta factors

geetu040 committed:
register decay factors as buffers

geetu040 committed:
remove unused declarations of decay factors

geetu040 committed:
update config for alpha/beta factors

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
run modular

geetu040 committed:
remove head_dim in tests

geetu040 commented:
@Cyrilvallez this is up for review again!

Cyrilvallez commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,680 @@
+                # inter: Q @ ( K.T @ V ) -> Q * KV
+                attn_output_inter = torch.matmul(current_query_states * current_query_decay, attn_weights_inter)
+
+                # final attention output
+                current_attn_output = attn_output_inter + attn_output_intra
+                attn_output.append(current_attn_output)
+
+                # cacluate attn_weights_inter for next block or cache
+                next_attn_weights_inter = torch.matmul(
+                    (current_key_states * current_key_decay).transpose(-1, -2), current_value_states
+                )
+                attn_weights_inter = attn_weights_inter * block_decay + next_attn_weights_inter
+
+        else:
+            ratio = torch.exp(-self.slope_rate)
+            attn_output = []
+            for i in range(seq_len):
+                current_query_states = query_states[:, :, i : i + 1]
+                current_key_states = key_states[:, :, i : i + 1]
+                current_value_states = value_states[:, :, i : i + 1]
+
+                current_attn_weights_inter = torch.matmul(current_key_states.transpose(-1, -2), current_value_states)
+                attn_weights_inter = ratio * attn_weights_inter + current_attn_weights_inter
+                current_attn_output = torch.matmul(current_query_states, attn_weights_inter)
+
+                attn_output.append(current_attn_output)
+
+        # concatenate attention outputs over all blocks
+        attn_output = torch.cat(attn_output, dim=-2)
+
+        # final output projection
+        attn_output = attn_output.transpose(1, 2)
+        attn_output = attn_output.reshape(batch_size, seq_len, self.num_attention_heads * self.head_dim)
+        attn_output = self.norm(attn_output)
+        attn_output = F.sigmoid(self.output_gate(hidden_states)) * attn_output
+        attn_output = self.out_proj(attn_output)
+
+        # update cache
+        if past_key_value is not None:
+            past_key_value.set_linear_cache(self.layer_idx, attn_weights_inter)
```
Here I'm concerned because it looks like we are not gonna get the same result with or without cache... Also, there should be some way to vectorize the 2 for loops no?

geetu040 replied:
> there should be some way to vectorize the 2 for loops no?

In the original code, there was a loop which I initially removed in [this commit](https://github.com/huggingface/transformers/pull/35831/commits/519eda3d2f5b2c1400095c3d10695043878f3c0d), but later reintroduced in [this commit](https://github.com/huggingface/transformers/pull/35831/commits/64f6e22a0a31a794127fa3309cb00db675af1fee).

The reason I brought the loop back is that I didn't observe any performance improvement after removing it. Without the loop, we still needed to create padding tensors, and a loop was still necessary to compute `attn_weights_inter` [here](https://github.com/huggingface/transformers/blob/519eda3d2f5b2c1400095c3d10695043878f3c0d/src/transformers/models/minimax_text_01/modeling_minimax_text_01.py#L169). Additionally, handling the decay factors becomes slightly more complex in the vectorized form.

That said, I'm fine with either implementation. I personally prefer the vectorized version, but since we still need a loop to compute `attn_weights_inter`, it somewhat defeats the purpose of removing the original loop.

For reference, the loop-free vectorized attention implementation can be found here: [MiniMaxText01LightningAttention](https://github.com/huggingface/transformers/blob/519eda3d2f5b2c1400095c3d10695043878f3c0d/src/transformers/models/minimax_text_01/modeling_minimax_text_01.py#L107).

geetu040 replied:
> I'm concerned because it looks like we are not gonna get the same result with or without cache.

Yes, that's correct—we do get different results when using the cache versus not using it. I looked into how the logits differ, and the discrepancy increases as the network goes deeper.

Cyrilvallez replied:
It should not be the case in general. Is it somehow expected from the original implementation? 🤔

geetu040 replied:
yes, this comes from the original implementation

Cyrilvallez commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,680 @@
+
+
+class MiniMaxModel(MixtralModel):
+    def __init__(self, config: MiniMaxConfig):
+        super().__init__(config)
+        self.layers = nn.ModuleList(
+            [MiniMaxDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
```
Should not be needed!

Cyrilvallez commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,680 @@
+    _supports_static_cache = False
+    _supports_quantized_cache = False
+
+    # TODO: remove init
+    def __init__(self, config):
+        super().__init__(config)
+        self.model = MiniMaxModel(config)
```
Indeed, this is not needed!

Cyrilvallez commented on src/transformers/utils/fx.py:
```diff
@@ -150,6 +150,7 @@ def _generate_supported_model_class_names(
     "megatron-bert",
     "mistral",
     "mixtral",
+    "minimax",
```
To remove!

Cyrilvallez commented on src/transformers/generation/utils.py:
```diff
@@ -3532,6 +3533,7 @@ def _beam_search(
                         "xlnet",
                         "cpm",
                         "jamba",
+                        "minimax",
```
This should be gone if you rebase

Cyrilvallez reviewed (commented):
Hey! Here are a few comments! 🤗 Very clean overall, but I'm a bit concerned by the fact that it looks like results will be different with or without cache in the LightningAttention module, could you maybe explain a bit?

geetu040 committed:
Merge branch main into minimax-text-01 (resolve conflicts)

geetu040 committed:
remove minimax from fx.py

geetu040 committed:
remove stuff that is not really needed

geetu040 committed:
update __init__

geetu040 committed:
update qkv torch.split

Co-authored-by: Cyril Vallez <cyril.vallez@gmail.com>

geetu040 committed:
fix qkv torch.split

geetu040 committed:
quality fixes

geetu040 committed:
remove mistakenly added dummy

geetu040 committed:
purge unused ModelTester code

geetu040 committed:
fix-copies

geetu040 commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,680 @@
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (`int`, *optional*, defaults to 32):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        num_key_value_heads (`int`, *optional*, defaults to 8):
+            This is the number of key_value heads that should be used to implement Grouped Query Attention. If
+            `num_key_value_heads=num_attention_heads`, the model will use Multi Head Attention (MHA), if
+            `num_key_value_heads=1` the model will use Multi Query Attention (MQA) otherwise GQA is used. When
+            converting a multi-head checkpoint to a GQA checkpoint, each group key and value head should be constructed
+            by meanpooling all the original heads within that group. For more details checkout [this
+            paper](https://arxiv.org/pdf/2305.13245.pdf). If it is not specified, will default to `8`.
```
please help me understand how is it not used in [MiniMaxAttention](https://github.com/huggingface/transformers/blob/8b00381ca6c8c615a30807ea7431ed269959cb99/src/transformers/models/minimax/modeling_minimax.py#L353)

ArthurZucker commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,678 @@
+                end_idx = min(start_idx + self.block_size, seq_len)
+                current_block_size = end_idx - start_idx
+
+                current_query_states = query_states[:, :, start_idx:end_idx].contiguous()
+                current_key_states = key_states[:, :, start_idx:end_idx].contiguous()
+                current_value_states = value_states[:, :, start_idx:end_idx].contiguous()
+
```
would calling query_states.contiguous() not ensure no strides on the slices?

geetu040 replied:
the `query_states` received here are not `contiguous`, I was applying `contiguous` on `current_query_states` because it was done in the source code but upon benchmarking, the overhead during `contiguous` does give a better trade-off, so I'll remove it

ArthurZucker commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,678 @@
+        residual = hidden_states
+
+        hidden_states = self.input_layernorm(hidden_states)
+        if self.postnorm:
```
to remove if not used on the released checkpoint!

ArthurZucker commented on tests/models/minimax/test_modeling_minimax.py:
```diff
@@ -0,0 +1,460 @@
+@require_torch
+@require_torch_accelerator
+@slow
+class MiniMaxIntegrationTest(unittest.TestCase):
```
we need a generation test here! 🤗

geetu040 replied:
added

ArthurZucker reviewed (approved):
Very nice! missing and integration test for generation otherwise LGTM. 
Great doc and great work standardizing!

geetu040 committed:
Merge branch "main"

geetu040 committed:
run fix-copies

geetu040 committed:
fix head_dim

geetu040 committed:
write cache formatting tests

geetu040 committed:
remove postnorm

geetu040 committed:
avoid contiguous in attention current states

geetu040 committed:
update expected_slice

geetu040 committed:
add generation test for integration

geetu040 committed:
fix dtype in generation test

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
update authors

geetu040 commented:
@Cyrilvallez, I have addressed the review comments.

How do you suggest we go with the checkpoint at [MiniMaxAI/MiniMax-Text-01](https://huggingface.co/MiniMaxAI/MiniMax-Text-01)?
The original safetensors can be used without any changes in the keys of state_dict, only the config is slightly different. We removed the `postnorm` and renamed attributes like `layernorm_full_attention_alpha` to `full_attn_alpha_factor` and the model architecture is `MiniMaxForCausalLM` instead of `MiniMaxText01ForCausalLM`.
Do we need a new repo like `MiniMaxAI/MiniMax-hf` or can the config be updated in the original repo?

geetu040 commented:
> Hi, @geetu040 @Shakib-IO . Thank you so much for all your hard work! Just wondering how things are going with the progress. Is there anything I can do to help move the PR forward?

Hi @sriting, thanks for checking in! The PR is complete and currently waiting on a review from @Cyrilvallez. In the meantime, I’ll update it with the latest from the main branch.

Since you're on the MiniMax team, I’ll give you write access in case you’d like to make any changes directly. Otherwise, we’d really appreciate your review!

ArthurZucker commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,668 @@
+            ratio = torch.exp(-self.slope_rate)
+            attn_output = []
+            for i in range(seq_len):
+                current_query_states = query_states[:, :, i : i + 1]
+                current_key_states = key_states[:, :, i : i + 1]
+                current_value_states = value_states[:, :, i : i + 1]
+
+                current_attn_weights_inter = torch.matmul(current_key_states.transpose(-1, -2), current_value_states)
+                attn_weights_inter = ratio * attn_weights_inter + current_attn_weights_inter
+                current_attn_output = torch.matmul(current_query_states, attn_weights_inter)
```
I am guesssing this would be optimized for flash attn var len?

ArthurZucker commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,668 @@
+            else:
+                input_attention_mask = causal_mask
+
+            if self.gradient_checkpointing and self.training:
```
we have `GradientCheckpointingLayer` for the `DecoderLayer` now that can just remove this!

ArthurZucker commented on src/transformers/models/minimax/modular_minimax.py:
```diff
@@ -0,0 +1,668 @@
+    pass
+
+
+class MiniMaxDecoderLayer(MixtralDecoderLayer):
```
```suggestion
class MiniMaxDecoderLayer(MixtralDecoderLayer, GradientCheckpointingLayer):
```

ArthurZucker reviewed (approved):
Let's go! 🚀 can you just rebase to main we had a few changes!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35831). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker commented:
Do you want me to commit directly to help get this merged?

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 commented:
@qscqesze Thanks for testing this out!

This could be happening because of minor difference in naming the config attributes

can you trying doing this with the config
```py
config = AutoConfig.from_pretrained(
    MODEL_PATH,
    trust_remote_code=True,
)

config = config.to_dict()
config['full_attn_alpha_factor'] = config['layernorm_full_attention_alpha']
config['full_attn_beta_factor'] = config['layernorm_full_attention_beta']
config['linear_attn_alpha_factor'] = config['layernorm_linear_attention_alpha']
config['linear_attn_beta_factor'] = config['layernorm_linear_attention_beta']
config['mlp_alpha_factor'] = config['layernorm_mlp_alpha']
config['mlp_beta_factor'] = config['layernorm_mlp_beta']

config = MiniMaxConfig(**config)

quantized_model = MiniMaxForCausalLM.from_pretrained(
    MODEL_PATH,
    config=config,
    torch_dtype="bfloat16",
    device_map="auto",
    quantization_config=quantization_config,
    attn_implementation="flash_attention_2",
    trust_remote_code=True,
)
```

Also, don't use the code from latest commit of this branch, I am fixing some bugs here, please use the previous commit https://github.com/huggingface/transformers/pull/35831/commits/9c5397d3220bfeb7abfbceb40e3a96f2aefa71de code

geetu040 committed:
update with changes in main

geetu040 committed:
update graident checkpointing and minor fixes

geetu040 committed:
fix mutable attn_type_list

geetu040 committed:
rename: attn_type -> layer_type

geetu040 commented:
@ArthurZucker please review the PR, if everything looks fine, we can move to handling the checkpoints on huggingface.

qscqesze commented:
Hello. I'm curious—why do we need to change attn_type_list to layer_type_list? I checked the config file at [config.json](https://huggingface.co/MiniMaxAI/MiniMax-Text-01/blob/main/config.json), and it actually doesn't contain a layer_type_list configuration.

When I use the latest code, I get the following error:


```
You are using a model of type minimax_text_01 to instantiate a model of type MiniMaxText01. This is not supported for all configurations of models and can yield errors.
You are using a model of type mixtral to instantiate a model of type MiniMaxText01. This is not supported for all configurations of models and can yield errors.
Traceback (most recent call last):
  File "/home/qingjun/vllm_test/huggingface_test.py", line 62, in <module>
    quantized_model = MiniMaxForCausalLM.from_pretrained(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/qingjun/transformers/src/transformers/modeling_utils.py", line 314, in _wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/qingjun/transformers/src/transformers/modeling_utils.py", line 4622, in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/qingjun/transformers/src/transformers/models/minimax/modeling_minimax.py", line 883, in __init__
    self.model = MiniMaxModel(config)
                 ^^^^^^^^^^^^^^^^^^^^
  File "/home/qingjun/transformers/src/transformers/models/minimax/modeling_minimax.py", line 657, in __init__
    [MiniMaxDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/qingjun/transformers/src/transformers/models/minimax/modeling_minimax.py", line 500, in __init__
    self.layer_type = config.layer_type_list[layer_idx]
                      ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
IndexError: list index out of range
```

但如果我修改

`config['layer_type_list'] = config['attn_type_list']`

就可以正常运行。

geetu040 committed:
Merge branch "main" (resolve conflicts)

geetu040 commented:
@qscqesze We actually handle this by creating another repository on huggingface, in this case something like `MiniMaxAI/MiniMax-Text-01-hf` which works with the implementation in transformers

ArthurZucker reviewed (approved):
My last comment then let's merge! cc @LysandreJik to merge once this is fixed

qscqesze commented:
@ArthurZucker @geetu040 Thank you both!
﻿
That makes perfect sense regarding the need for a separate Hugging Face repo and the renaming to align with the new `layer_type` standards.

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 committed:
update for layer_types

geetu040 committed:
update integration tests

geetu040 committed:
update checkpoint

geetu040 committed:
clean overview in docs

geetu040 committed:
Merge branch 'main' into minimax-text-01

geetu040 commented:
> My last comment then let's merge! cc @LysandreJik to merge once this is fixed

Hi @LysandreJik, This PR is ready for merge!

These checkpoints for integration tests should also be moved from [`geetu040/MiniMax-tiny`](https://huggingface.co/geetu040/MiniMax-tiny) to [`hf-internal-testing/MiniMax-tiny`](https://huggingface.co/hf-internal-testing/MiniMax-tiny).

LysandreJik commented:
Hey @geetu040, I mirrored your tiny checkpoint here: https://huggingface.co/hf-internal-testing/MiniMax-tiny/tree/main

I'm merging the PR! 

Thanks for your contribution 🤗

hadipash commented on src/transformers/models/minimax/configuration_minimax.py:
```diff
@@ -0,0 +1,230 @@
+    This is the configuration class to store the configuration of a [`MiniMaxModel`]. It is used to instantiate an
+    MiniMax model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the MiniMax.
+
+    [MiniMaxAI/MiniMax-Text-01-hf](https://huggingface.co/MiniMaxAI/MiniMax-Text-01-hf)
```
When will this repo be available?

geetu040 replied:
I've also opened a PR on the hf model repo that might be helpful:
[https://huggingface.co/MiniMaxAI/MiniMax-Text-01/discussions/39](https://huggingface.co/MiniMaxAI/MiniMax-Text-01/discussions/39)

geetu040 replied:
@hadipash you should now be able to see the repo at [MiniMaxAI/MiniMax-Text-01-hf](https://huggingface.co/MiniMaxAI/MiniMax-Text-01-hf)
